src/main/java/de/tum/in/www1/artemis/versioning/VersionRequestMappingHandlerMapping.java

```diff
@@ -0,0 +1,94 @@
+import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping;
+
+import de.tum.in.www1.artemis.exception.ApiVersionAnnotationMismatchException;
+
+/**
+ * This class is responsible for integrating the API versions into the request mapping.
+ */
+public class VersionRequestMappingHandlerMapping extends RequestMappingHandlerMapping {
+
+    private final String pathPrefixSegment;
+
+    private final String versionPrefix;
+
+    private final List<Integer> apiVersions;
+
+    private final int latestVersion;
+
+    /**
+     * Creates a new VersionRequestMappingHandlerMapping.
+     *
+     * @param pathPrefixSegment The full path segment on the beginning of endpoint path
+     * @param versionPrefix     The prefix for the version segment
+     */
+    public VersionRequestMappingHandlerMapping(List<Integer> apiVersions, String pathPrefixSegment, String versionPrefix) {
+        this.apiVersions = apiVersions;
+        this.latestVersion = this.apiVersions.get(this.apiVersions.size() - 1);
+        this.pathPrefixSegment = pathPrefixSegment;
+        this.versionPrefix = versionPrefix;
+    }
+
+    /**
+     * This method is called by the Spring framework to get the mapping for a specific endpoint method containing all versions.
+     *
+     * @param method      the method to provide a mapping for
+     * @param handlerType the handler type, not relevant to us
+     * @return the mapping for the method
+     */
+    @Override
+    protected RequestMappingInfo getMappingForMethod(@NotNull Method method, @NotNull Class<?> handlerType) {
+        RequestMappingInfo info = super.getMappingForMethod(method, handlerType);
+        if (info == null) {
+            return null;
+        }
+        // We don't handle endpoints we didn't create
+        if (!method.getDeclaringClass().getPackageName().contains("de.tum.in.www1.artemis")) {
+            return info;
+        }
+        // We don't handle endpoints that are explicitly ignored
+        var ignoreGlobalMappingAnnotation = method.getAnnotation(IgnoreGlobalMapping.class);
+        if (ignoreGlobalMappingAnnotation != null && ignoreGlobalMappingAnnotation.ignorePaths()) {
+            return info;
+        }
+        VersionRanges versionRangesAnnotation = AnnotationUtils.findAnnotation(method, VersionRanges.class);
+        VersionRange versionRangeAnnotation = AnnotationUtils.findAnnotation(method, VersionRange.class);
+        if (versionRangesAnnotation != null && versionRangeAnnotation != null) {
+            throw new ApiVersionAnnotationMismatchException();
+        }
+        if (versionRangesAnnotation != null) {
+            return createApiVersionInfo(new VersionRangesRequestCondition(apiVersions, versionRangesAnnotation.value())).combine(info);
+        }
+        if (versionRangeAnnotation != null) {
+            return createApiVersionInfo(new VersionRangesRequestCondition(apiVersions, versionRangeAnnotation)).combine(info);
+        }
```

Comment:
Not really. It would get unnecessarily complicated. If I wanted to use a ternary operator, which would be necessary here, the parameters would have to be of the same type. But to leave it empty, I would have to use a set; the values of the annotations are given as arrays. That's incompatible.

I could also give the whole condition in a ternary operation, but that would not make it more readable.